README.md
# discord_form_flutter

A simple, clean designed form. Similar to Discord's. ⚡

<img src="screenshot.png" alt="screenshot" width="300"/>


## Getting Started

### Add this to your package's pubspec.yaml file:

```yaml
dependencies:
  discord_form_flutter: ^0.0.1
```

### Import the package

```dart
import 'package:discord_form_flutter/discord_form_flutter.dart';
```

## How to Use

Call the `DiscordForm` widget and pass the reqiured parameters.
You can use it as a complete page. ✔

```
final _emailController = TextEditingController();
final _formKey = GlobalKey<FormState>();

@override
Widget build(BuildContext context) {
  return MaterialApp(
    home: DiscordForm(
    formKey: _formKey,
    children: [
      DiscordFormField(
        hint: 'Email',
        icon: Icon(Icons.email),
        onChanged: (value) {
              // handle textfield changes
        },
        controller: _emailController,
        ),
        SizedBox(height: 35),
        DiscordButton(
          onPressed: () {
              // handle button press
          },
          text: Text(
            'Login',
          ),
        ),
    ],
   ),
  );
}
```

## Example

You can find an example [here](https://github.com/AymanProjects/discord_form_flutter/blob/master/example/lib/main.dart)



CHANGELOG.md
## [0.0.1+1] - 2021-02-25

* Fixed documentation and formatting

## [0.0.1] - 2021-02-24

* The initial version of discord_form_flutter

lib/src/discordForm.dart
import 'package:flutter/material.dart';
import 'defaultTheme.dart';

class DiscordForm extends StatelessWidget {
  /// Pass a [FormState] key to use it for validation.
  /// All children of this form will be validated automatically
  final GlobalKey<FormState> formKey;

  /// The children of te form can be any widget
  /// but mostly you will only need [DiscordFormField] and [DiscordButton]
  final List<Widget> children;

  /// How the validation should behave.
  /// default is [AutovalidateMode.disabled]
  /// which means it will only validate when
  /// [formKey.currentState.validate()] is called
  final AutovalidateMode autovalidateMode;

  /// Setting this to [EdgeInsets.zero] will
  /// cause the form to expand and fit it's parent.
  /// default is horizontal: 40.0 and vertical: 30.0
  final Padding outerPadding;

  /// Limit the maximum width for the form.
  /// default is 450.0
  final double maxWidth;

  /// by default it's false, but
  /// if set to true then the theme used on the form
  /// and it's children will be derived from you app theme
  final bool useTheAppTheme;

  /// Simple and clean form.
  /// In the chilren, you can pass any widget but
  /// preferably [DiscordFormField] and [DiscordButton]
  DiscordForm({
    @required this.formKey,
    @required this.children,
    this.autovalidateMode = AutovalidateMode.disabled,
    this.outerPadding,
    this.maxWidth = 450.0,
    this.useTheAppTheme = false,
  })  : assert(formKey != null),
        assert(children != null),
        assert(autovalidateMode != null),
        assert(maxWidth != null),
        assert(useTheAppTheme != null);

  @override
  Widget build(BuildContext context) {
    return Theme(
      /// if useTheAppTheme is false, then a pre-made theme will be applied
      data: useTheAppTheme ? Theme.of(context) : defaultTheme,
      child: Scaffold(
        body: Center(
          /// wrap children in SingleChildScrollView to avoid overflowing
          child: SingleChildScrollView(
            child: Container(
              constraints: BoxConstraints(maxWidth: maxWidth),
              width: double.infinity,
              padding: outerPadding ??
                  const EdgeInsets.symmetric(horizontal: 40.0, vertical: 30.0),
              child: Form(
                autovalidateMode: autovalidateMode,
                key: formKey,
                child: Column(
                  children: children,
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


lib/src/defaultTheme.dart
import 'package:flutter/material.dart';

/// default color palette
final Color blue = Color(0xff7289da);
final Color darkBlue = Color(0xff353840);
final Color darkerBlue = Color(0xff323238);

/// A default theme in case the form them was not set
final ThemeData defaultTheme = ThemeData(
  /// form field theme
  inputDecorationTheme: InputDecorationTheme(
    fillColor: darkerBlue,
  ),
  textTheme: TextTheme(
    /// The form field text style
    subtitle1: TextStyle(color: Colors.white),
  ),
  scaffoldBackgroundColor: darkBlue,
  primaryColor: blue,
  accentColor: Colors.white.withOpacity(0.5),
  cursorColor: blue,
  brightness: Brightness.dark,

  /// to set the text color inside the button
  buttonTheme: ButtonThemeData(
    textTheme: ButtonTextTheme.accent,
    colorScheme: ColorScheme.dark(
      /// button text color
      secondary: Colors.white,
    ),
  ),
);


lib/src/discordButton.dart
import 'package:flutter/material.dart';

class DiscordButton extends StatelessWidget {
  /// The [Text] widget inside the button.
  /// Can't be null
  /// Default is white
  final Text text;

  /// Pass a function that you would like to trigger
  /// whenever the button is pressed
  final VoidCallback onPressed;

  /// [isLoading] is used to control the button ability to be pressed,
  /// it's [false] by default, but if you set it to [true]
  /// it will be disabled and will show a [CircularProgressIndicator]
  /// in the center of the button
  final bool isLoading;

  /// Use this widget as a button in [DiscordForm].
  /// text and isLoading must not be null
  DiscordButton({
    @required this.text,
    this.onPressed,
    this.isLoading = false,
  })  : assert(text != null),
        assert(isLoading != null);

  @override
  Widget build(BuildContext context) {
    return Container(
      /// expand to available width
      width: double.infinity,
      child: FlatButton(
        shape: RoundedRectangleBorder(
          borderRadius: const BorderRadius.all(Radius.circular(4)),
        ),

        /// control the button color using [Theme.primaryColor]
        color: Theme.of(context).primaryColor,
        disabledColor: Theme.of(context).primaryColor.withOpacity(0.3),

        /// disable the button if isLoading is true
        onPressed: isLoading ? null : onPressed,
        child: Padding(
          padding: const EdgeInsets.all(12.0),
          child: isLoading
              ? Container(
                  height: 20.0,
                  width: 20.0,

                  /// Indicator color will be the same as [Theme.accentColor]
                  child: CircularProgressIndicator(
                    strokeWidth: 2,
                  ),
                )
              : text,
        ),
      ),
    );
  }
}


lib/src/discordFormField.dart
import 'package:flutter/material.dart';

class DiscordFormField extends StatelessWidget {
  /// Display a text for as a hint inside the  form field.
  /// Style can be changed by overriding the [InputDecorationTheme]
  /// of your app theme
  final String hint;

  /// Pass a function that validates the field input.
  /// You should return a string in case of error,
  /// and return null if the input is valid.
  ///
  /// If you set it to null, no validations will apply
  final String Function(String) validator;

  /// Pass a function that you would like to trigger
  /// whenever the field is modified
  final Function(String) onChanged;

  /// If set to true, text will be obscured
  final bool isPassword;

  /// If set to true, the keyboard type will
  /// be [TextInputType.emailAddress]
  final bool isEmail;

  /// The suffix icon inside the field.
  /// Set it to null if you don't want to show
  /// an icon
  final Icon icon;

  /// A controller to store the state of the field.
  /// Can't be null
  final TextEditingController controller;

  /// Use this widget as a textField in [DiscordForm].
  DiscordFormField({
    this.hint = 'Field',
    this.validator,
    @required this.onChanged,
    this.isPassword = false,
    this.isEmail = false,
    this.icon,
    this.controller,
  })  : assert(hint != null),
        assert(onChanged != null),
        assert(isPassword != null),
        assert(isEmail != null),
        assert(controller != null);

  @override
  Widget build(BuildContext context) {
    return Theme(
      data: Theme.of(context).copyWith(accentColor: Colors.white),
      child: TextFormField(
        controller: controller,

        /// if isEmail is false, show a regular keyboard
        keyboardType: isEmail ? TextInputType.emailAddress : TextInputType.text,
        obscureText: isPassword,
        onChanged: onChanged,
        validator: validator,
        decoration: InputDecoration(
          hintText: hint,
          suffixIcon: icon,
          contentPadding: const EdgeInsets.all(12.0),

          /// The fill color is [Theme.inputDecorationTheme.fillColor]
          filled: true,
          enabledBorder: _border(),
          focusedBorder: _border(color: Theme.of(context).primaryColor),
          errorBorder: _border(color: Theme.of(context).errorColor),
          focusedErrorBorder: _border(color: Theme.of(context).errorColor),
        ),
      ),
    );
  }

  OutlineInputBorder _border({Color color}) {
    return OutlineInputBorder(
      borderSide: BorderSide(
        color: color ?? Colors.black,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:discord_form_flutter/discord_form_flutter.dart';

main() {
  runApp(
    MaterialApp(
      home: App(),
    ),
  );
}

class App extends StatefulWidget {
  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<App> {
  /// to preserve the state of form fields, you have to
  /// create a controller for each.
  final emailController = TextEditingController();
  final passwordController = TextEditingController();

  /// a form key, used for validation
  final _formKey = GlobalKey<FormState>();

  /// to store the entered email
  String _email;

  /// to store the entered password
  String _password;

  /// control when the button should be loading
  bool _loading = false;

  @override
  Widget build(BuildContext context) {
    return DiscordForm(
      formKey: _formKey,
      children: [
        DiscordFormField(
          /// to show an email keybaord, set this to true
          isEmail: true,
          hint: 'Email',
          icon: Icon(Icons.email),
          onChanged: (value) {
            /// update our local variable
            _email = value;
          },
          controller: emailController,

          /// return null when the used input is valid
          validator: (value) => value.isEmpty ? 'can\'t be empty' : null,
        ),

        /// add vertical spacing between fields
        SizedBox(height: 25),

        ///
        DiscordFormField(
          /// to hide the password, set this to true
          isPassword: true,
          hint: 'Password',
          icon: Icon(Icons.lock),
          onChanged: (value) {
            _password = value;
          },
          controller: passwordController,

          /// return null when the used input is valid
          validator: (value) =>
              value.length < 6 ? 'minimum 6 characters' : null,
        ),
        SizedBox(height: 35),
        DiscordButton(
          isLoading: _loading,
          onPressed: () {
            if (_formKey.currentState.validate()) {
              print('logged in succesfully!'
                  ' Email: $_email, Password: $_password');
            }
          },
          text: Text(
            'Login',
            style: TextStyle(fontSize: 15.0),
          ),
        ),
      ],
    );
  }
}
